src/mpi_t/qmpi_register.c

```diff
@@ -0,0 +1,402 @@
+
+/*
+ * Copyright (C) by Argonne National Laboratory
+ *     See COPYRIGHT in top-level directory
+ */
+
+#include "mpiimpl.h"
+
+#ifdef ENABLE_QMPI
+
+#include "qmpi_register.h"
+#include <execinfo.h>
+#define BACKTRACE_SIZE 5
+
+/*
+=== BEGIN_MPI_T_CVAR_INFO_BLOCK ===
+
+categories:
+    - name        : TOOLS
+      description : cvars that control tools connected to MPICH
+
+cvars:
+    - name        : MPIR_CVAR_QMPI_TOOL_LIST
+      alias       : QMPI_TOOL_LIST
+      category    : TOOLS
+      type        : string
+      default     : NULL
+      class       : none
+      verbosity   : MPI_T_VERBOSITY_USER_BASIC
+      scope       : MPI_T_SCOPE_ALL_EQ
+      description : >-
+        Set the number and order of QMPI tools to be loaded by the MPI library when it is
+        initialized.
+
+=== END_MPI_T_CVAR_INFO_BLOCK ===
+*/
+
+typedef void (*generic_mpi_func) (void);
+
+void (**MPIR_QMPI_pointers) (void);
+void **MPIR_QMPI_storage;
+void (**MPIR_QMPI_tool_init_callbacks) (int);
+int MPIR_QMPI_num_tools = 0;
+char **MPIR_QMPI_tool_names;
+void (**MPIR_QMPI_first_fn_ptrs) (void);
+int *MPIR_QMPI_first_tool_ids;
+
+/* Cache the first function and tool ID in the chain of QMPI tools to optimize the lookup in all of
+ * the internal MPI_* functions. */
+int MPII_qmpi_stash_first_tools(void)
```

Comment:
Depend on whether my understanding is correct, but I think this can be a static function.

Reply:
It needs to be called by all of the wrapper functions (e.g., `MPI_Send`) so I don't think it can (unless it's a header).